ESCOMP/CTSM
# mksurfdata: Rework mapping of landunits to handle coastal areas more rigorously

### Description of changes

This PR reworks the mapping of landunit areas and the determination of wetland areas according to the plan laid out in #1716. (Note that the final implementation differs from the initial comment in #1716, as discussed in follow-up comments in that issue.) See the discussion there as well as the design document in this PR for details.

### Specific notes

Contributors other than yourself, if any: Various discussions with others, especially @ekluzek and @slevisconsulting 

CTSM Issues Fixed (include github issue #):
- Resolves #1716 
- Resolves #1930 

Are answers expected to change (and if so in what way)? Areas will change on surface datasets

Any User Interface Changes (namelist or namelist defaults changes)? No

Testing performed, if any:

(1) Compared 1850 f09 surface dataset before and after these changes (and after each individual change from the individual commits on this branch)

(2) By hard-coding landunit areas before the call to `normalize_and_check_landuse`, I performed various "manual unit tests" of the normalization code, using diffs like this:

``` diff
diff --git a/tools/mksurfdata_esmf/src/mksurfdata.F90 b/tools/mksurfdata_esmf/src/mksurfdata.F90
index eb91be30c..0f347f8c0 100644
--- a/tools/mksurfdata_esmf/src/mksurfdata.F90
+++ b/tools/mksurfdata_esmf/src/mksurfdata.F90
@@ -593,6 +593,17 @@ program mksurfdata
   ! Perform other normalizations
   ! -----------------------------------
 
+  do n = 1, lsize_o
+     pctlnd_pft(n) = 0._r8
+
+     call pctnatpft(n)%set_pct_l2g(0._r8)
+     call pctcft(n)%set_pct_l2g(0._r8)
+     pctlak(n) = 0._r8
+     pcturb(n) = 0._r8
+     pctgla(n) = 0._r8
+     pctwet(n) = 0._r8
+  end do
+
   ! Normalize land use and make sure things add up to 100% as well as
   ! checking that things are as they should be.
   call normalize_and_check_landuse(lsize_o)
```

In the following, any landunit not explicitly mentioned had 0% area:

(a) With pctlnd_pft = 0% and all initial landunit areas 0%: wetland area = 100%

(b) With pctlnd_pft = 0% and initial glacier area 1%: glacier area = 100%

(c) With pctlnd_pft > 0% and all initial landunit areas 0%: natural vegetated area = 100%

(d) With pctlnd_pft = 40%, initial crop area 20%, natural vegetated area 10%: crop area = 50%, natural vegetated area = 50%

(e) With pctlnd_pft = 40%, initial crop area 20%, natural vegetated area 10%, glacier area 10%: crop area = 50%, natural vegetated area = 25%, glacier area = 25%

(f) With pctlnd_pft = 40%, initial crop area 20%, natural vegetated area 10%, glacier area 15%: crop area = 50%, natural vegetated area = 12.5%, glacier area = 37.5%

(g) With pctlnd_pft = 40%, initial crop area 20%, natural vegetated area 10%, glacier area 20%: crop area = 50%, glacier area = 50%

(h) With pctlnd_pft = 40%, initial crop area 20%, natural vegetated area 10%, glacier area 30%: crop area = 40%, glacier area = 60%

(i) With pctlnd_pft = 40%, initial crop area 0%, natural vegetated area 40%, glacier area 40%: glacier area = 100%

(j) With pctlnd_pft = 2%, initial natural vegetated area 1%, glacier area 1%: natural vegetated area = 50%, glacier area = 50%

(k) With pctlnd_pft = 2%, initial natural vegetated area 0%, glacier area 1%: natural vegetated area = 50%, glacier area = 50%

(l) With pctlnd_pft = 2%, initial natural vegetated area 2%, glacier area 1%: natural vegetated area = 50%, glacier area = 50%
